pkg/networkplugin-syncer/handlers/ovn/infrastructure.go

```diff
@@ -0,0 +1,80 @@
+package ovn
+
+import (
+	goovn "github.com/ebay/go-ovn"
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+const (
+	submarinerLogicalRouter = "submariner_router"
+	ovnClusterRouter        = "ovn_cluster_router"
+)
+
+func (ovn *SyncHandler) ensureSubmarinerInfra() error {
+	if err := ovn.ensureSubmarinerJoinSwitch(); err != nil {
+		return err
+	}
+
+	if err := ovn.ensureSubmarinerRouter(); err != nil {
+		return err
+	}
+
+	if err := ovn.connectOvnClusterRouterToSubm(); err != nil {
+		return err
+	}
+
+	// At this point, we are missing the ovn_cluster_router policies and the
+	// local-to-remote & remote-to-local routes in submariner_router, that
+	// depends on endpoint details that we will receive via events
+
+	return nil
+}
+
+func (ovn *SyncHandler) ensureSubmarinerJoinSwitch() error {
+	lsCmd, err := ovn.nbdb.LSAdd(submarinerDownstreamSwitch)
+	if err == nil {
+		klog.Infof("Creating submariner switch %q", submarinerDownstreamSwitch)
+
+		err = ovn.nbdb.Execute(lsCmd)
+	}
+
+	if !errors.Is(err, goovn.ErrorExist) {
+		return errors.Wrapf(err, "Creating submariner switch %q", submarinerDownstreamSwitch)
+	}
+
+	return nil
+}
+
+func (ovn *SyncHandler) ensureSubmarinerRouter() error {
+	lrCmd, err := ovn.nbdb.LRAdd(submarinerLogicalRouter, nil)
+	if err == nil {
+		klog.Infof("Creating submariner router %q", submarinerLogicalRouter)
+
+		err = ovn.nbdb.Execute(lrCmd)
+	}
+
+	if !errors.Is(err, goovn.ErrorExist) {
+		return errors.Wrapf(err, "Creating submariner router %q", submarinerLogicalRouter)
+	}
+
+	// TODO: Improve this so we don't need to delete/recreate everytime
+	delOldRouterPort, _ := ovn.nbdb.LRPDel(submarinerLogicalRouter, submarinerDownstreamRPort)
+	_ = ovn.nbdb.Execute(delOldRouterPort)
+	delOldLSP, _ := ovn.nbdb.LSPDel(submarinerDownstreamSwPort)
+	_ = ovn.nbdb.Execute(delOldLSP)
```

Comment:
Yes, I need to make some improvements into the ovndb library to avoid this.

If we don't do it, there is the risk that the link cmd down there won't work.

That is why I added a TODO

Reply:
I extended the TODO explanation